### ISSUE TITLE
warn on casting integer division to a float

### ISSUE THREAD

**Issue by sourcefrog** [rust-lang/rust-clippy#8251]
### What it does

Maybe Clippy should warn about dividing two integers and then casting to a float? For example, `(4/3) as f64` probably does not do what the author intends. Clippy could suggest instead writing `4f64 / 3f64`.

This could perhaps be extended to division of integer variables or expressions, although perhaps in that case there might be more cases where it's intentional?

I'm not sure if this is a common-enough mistake to be worth a lint.

I read this post describing someone's experience using Rust for the first time and making this mistake: https://blog.singleton.io/posts/2022-01-02-raytracing-with-rust/#:~:text=let%20aspect_ratio%20%3D%20(800%20/%20600)%20as%20f64

### Lint Name

cast_integer_division_to_float

### Category

suspicious

### Advantage

This may be a bug: the division will be done in integers, when the user might intend to get a fractional result.

### Drawbacks

False positives? Perhaps bugs from this mistake will often be obvious?

### Example

```rust
let r = (800 / 600) as f64;
let a = 13;
let b = (a / 3) as f64;
```

Could be written as:

```rust
let r = 800.0 / 600.0;
let a = 13;
let b = (a as f64) / 13;
```

**Comment by giraffate** [comment]
FYI maybe a little different, but the [`integer_division`](https://rust-lang.github.io/rust-clippy/master/index.html#integer_division) might be useful in this case.

**Comment by sourcefrog** [comment]
Thanks @giraffate . It's probably at least useful as a guide for implementation.

I see integer division is classified as `restriction` which makes sense as it's a pretty common and legitimate thing to do, and although there's some risk of using it incorrectly I don't think many projects would want to always be warned about it.

Potentially, integer division that's then cast to a float is more likely to be an outright mistake, and may be something that's more reasonable to warn about by default?

I could have a go at implementing this but I'd appreciate any feedback on whether people think it would be reasonable to turn on by default, or how we should go about establishing that.

**Comment by sourcefrog** [comment]
I agree it's very reasonable that `3/4` is `restriction`. 

I am suggesting that perhaps `(3/4) as f64` should be `suspicious` or some other category that is "higher" than `restriction`.

**Comment by giraffate** [comment]
> I am suggesting that perhaps `(3/4) as f64` should be `suspicious` or some other category that is "higher" than restriction.

Oh, I see. It looks good.
